### ISSUE TITLE
Feature request: Profile-aware prompt router with ML-backed routing and ensemble fallback

### ISSUE THREAD

**Issue by zavera** [langchain-ai/langchain#38390]
## Summary

I am proposing a new router type for LangChain that uses classical ML to select among prompt variants based on a structured feature vector extracted from the input. This is distinct from existing semantic routers which use embedding similarity. The use case is structured computation pipelines where prompt performance varies predictably by input profile characteristics.

## Motivation

Existing LangChain routers (semantic router, LLM router) make routing decisions based on query semantics or ask an LLM to decide. For structured computation tasks, neither is ideal. The routing signal is not semantic similarity to example queries. It is a small set of boolean features characterizing the input payload structure.

For example: in tax computation, whether Schedule 1 is present or absent is the dominant predictor of which prompt formula produces the correct output. A KNN or Random Forest trained on a small labeled evaluation matrix routes with far higher accuracy than semantic similarity or LLM self-routing.

## Proposed Component

`ProfileAwareRouter` as a new router in `langchain.chains.router`

```python
from langchain.chains.router import ProfileAwareRouter
from langchain.chains.router.ml_routing import KNNRoutingModel

router = ProfileAwareRouter(
    destination_chains={
        "v1": llm_chain_v1,
        "v2": llm_chain_v2,
        "v3": llm_chain_v3,
    },
    feature_extractor=extract_profile_features,
    routing_model=KNNRoutingModel(n_neighbors=5),
    confidence_threshold=0.7,
    fallback_chain=ensemble_chain,  # used when confidence is below threshold
)

result = router.invoke({"payload": student_tax_payload})
```

## Architecture

**Feature extraction layer**
A user-defined callable that maps any input to a fixed-length feature vector. For boolean presence flags this is low-dimensional and stable. Keeps the routing model lightweight and interpretable.

**ML routing model**
Pluggable sklearn-compatible model. KNN for interpretability and small datasets. Random Forest for larger datasets and feature importance reporting. Trained offline on a labeled evaluation matrix where humans have marked which prompt produced the correct output per input profile.

**Confidence scoring**
KNN returns neighbor agreement as confidence. Random Forest returns class probability. When confidence is below threshold, route to fallback ensemble chain rather than hard routing.

**Ensemble fallback chain**
A chain that runs all destination chains and reconciles outputs using weighted voting or weighted median (for numeric outputs). Weights derived from per-cluster historical accuracy in the evaluation matrix.

**Feedback loop compatibility**
Router exposes a `record_outcome(input, routed_to, correct)` method so downstream verification signals can flow back and trigger retraining. Designed for pipelines where ground truth is verifiable after the fact.

## Relationship to Existing Routers

This complements rather than replaces existing routers:

| Router type        | Signal used               | Best for                     |
|--------------------|---------------------------|------------------------------|
| Semantic router    | Embedding similarity      | Open-ended query routing     |
| LLM router         | LLM judgment              | Complex routing logic        |
| ProfileAwareRouter | ML on structured features | Structured computation tasks |

## Statistical Validation Built In

The routing model exposes standard sklearn validation:
- Silhouette score for cluster quality
- Out-of-bag error for Random Forest
- Calibration curve for confidence score reliability
- Per-segment precision/recall breakdown
- Routing lift vs best single prompt baseline

## Contribution Interest

I am actively building this and would be happy to contribute a reference implementation. Interested in feedback on whether this fits within `langchain.chains.router` or belongs in a separate `langchain.routing` module, and whether the sklearn dependency should be optional.
